src/charm.py

```diff
@@ -1,117 +1,282 @@
+            old_csr = self._tls.get_relation_data_field(
+                f"csr-{hostname}",
+                tls_certificates_relation,  # type: ignore[arg-type]
+            )
+            if not old_csr:
+                continue
+            if JujuVersion.from_environ().has_secrets:
+                try:
+                    secret = self.model.get_secret(label=f"private-key-{hostname}")
+                    secret.remove_all_revisions()
+                except SecretNotFoundError:
+                    LOGGER.warning("Juju secret for %s already does not exist", hostname)
+                    continue
+            try:
+                self._tls.pop_relation_data_fields(
+                    [f"key-{hostname}", f"password-{hostname}"],
+                    tls_certificates_relation,  # type: ignore[arg-type]
+                )
+                peer_relation = self.model.get_relation("nginx-peers")  # type: ignore[arg-type]
+                self._tls.pop_relation_data_fields(
+                    [f"key-{hostname}", f"password-{hostname}"],
+                    peer_relation,  # type: ignore[arg-type]
+                )
+            except KeyError:
+                LOGGER.warning("Relation data for %s already does not exist", hostname)
+            self.certificates.request_certificate_revocation(
+                certificate_signing_request=old_csr.encode()
+            )
+
+    def _on_certificate_invalidated(self, event: CertificateInvalidatedEvent) -> None:
+        """Handle the TLS Certificate invalidation event.
+
+        Args:
+            event: The event that fires this method.
+        """
+        tls_certificates_relation = self._tls.get_tls_relation()
+        if not tls_certificates_relation:
+            self.unit.status = WaitingStatus("Waiting for certificates relation to be created")
+            event.defer()
+            return
+        if event.reason == "revoked":
+            hostname = self._tls.get_hostname_from_cert(event.certificate)
+            self._certificate_revoked([hostname])
+        if event.reason == "expired":
+            self._tls.certificate_expiring(event, self.certificates)
+        self.unit.status = MaintenanceStatus("Waiting for new certificate")
+
+    def _on_all_certificates_invalidated(self, _: AllCertificatesInvalidatedEvent) -> None:
+        """Handle the TLS Certificate relation broken event.
+
+        Args:
+            _: The event that fires this method.
+        """
+        tls_relation = self._tls.get_tls_relation()
+        if tls_relation:
+            tls_relation.data[self.app].clear()
+        if JujuVersion.from_environ().has_secrets:
+            hostname = self.get_hostname()
+            try:
+                secret = self.model.get_secret(label=f"private-key-{hostname}")
+                secret.remove_all_revisions()
+            except SecretNotFoundError:
+                LOGGER.warning("Juju secret for %s already does not exist", hostname)
```

Comment:
Just out of curiosity, may I ask when the secret may already be gone?

Reply:
Might be when it's manually deleted by the operator via `juju remove-secret`